src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/controllers/UserControllerTests.kt

```diff
@@ -0,0 +1,164 @@
+class UserControllerTests {
+  lateinit var userService: UserService
+  lateinit var requestBuilder: RequestBuilder
+  var endpoint = "/users/"
+  var id1 = "70b664a416135a6e967fadc6"
+
+  @BeforeEach
+  fun configureSystem() {
+    userService = mock(UserService::class.java)
+    val userController = UserController(userService)
+    val mockMvc = MockMvcBuilders.standaloneSetup(userController)
+      .build()
+    requestBuilder = RequestBuilder(mockMvc)
+  }
+
+  @Nested
+  @DisplayName("Tests for getUsers")
+  inner class GetUsers {
+    @Test
+    fun shouldReturnAllUsers() {
+      val user1 = User("id1", "username", "so secure", "test@test.com")
+      val user2 = User("id2", "sauce1", "123456", "fun@fun.com")
+      val users = listOf(user1, user2)
+      given(userService.getAll()).willReturn(users)
+      requestBuilder.runGetRequest(endpoint)
+        .andExpect(status().isOk)
+        .andExpect(jsonPath("$[0].username", equalTo(user1.username)))
+        .andExpect(jsonPath("$[0].password", equalTo(user1.password)))
+        .andExpect(jsonPath("$[0].email", equalTo(user1.email)))
+        .andExpect(jsonPath("$[1].username", equalTo(user2.username)))
+        .andExpect(jsonPath("$[1].password", equalTo(user2.password)))
+        .andExpect(jsonPath("$[1].email", equalTo(user2.email)))
+
+    }
+  }
+
+  @Nested
+  @DisplayName("Tests for getSingleUser")
+  inner class GetSingleUser {
+    @Test
+    fun shouldReturnUserSuccessfully() {
+      val user = User("id1", "username", "so secure", "test@test.com")
+      given(userService.getSingle(any())).willReturn(user)
+      endpoint += "$id1/"
+      requestBuilder.runGetRequest(endpoint)
+        .andExpect(status().isOk)
+        .andExpect(jsonPath("$.id", equalTo("id1")))
+        .andExpect(jsonPath("$.username", equalTo(user.username)))
+        .andExpect(jsonPath("$.password", equalTo(user.password)))
+        .andExpect(jsonPath("$.email", equalTo(user.email)))
+    }
+
+    @Test
+    fun shouldNotFindUser() {
+      given(userService.getSingle(any())).willReturn(null)
+      endpoint += "$id1/"
+      requestBuilder.runGetRequest(endpoint)
+        .andExpect(status().isNotFound)
+        .andExpect(jsonPath("$").doesNotExist())
+    }
+  }
+  
+  @Nested
+  @DisplayName("Tests for createUser")
+  inner class CreateUser {
+    @Test
+    fun shouldSuccessfullyCreateUser() {
+      val user = User("id1", "epicsauceXD", "123", "dude@something.com")
+      given(userService.create(any())).willReturn(user)
+      val userRequest = UserRequest("epicsauceXD", "123", "dude@something.com")
+      requestBuilder.runPostRequest(endpoint, requestToString(userRequest))
+        .andExpect(status().isCreated)
+        .andExpect(jsonPath("$.username", equalTo(user.username)))
+        .andExpect(jsonPath("$.password", equalTo(user.password)))
+        .andExpect(jsonPath("$.email", equalTo(user.email)))
```

Comment:
Should we test that "id" exists as well?

Reply:
(probably, I got lazy :P will update)